### PR TITLE
[GPU] Remove duplicated OpenCL kernel compilation on static model

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
 - *Remove duplicated kernels in internal kernel source map in kernels_cache*
 - *Replace unique id from processing idx to kernel_impl_param.hash()*

### Tickets:
 - *60354*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/post_input_reorder.cpp [hunk 1]
```diff
@@ -65,7 +65,6 @@ void post_input_reorder::run(program& p) {
                                       layout_format,
                                       input_layout.data_padding);
                 auto& reorder = add_reorder(p, input, node, current_layout);
-                reorder.set_unique_id();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/post_input_reorder.cpp:68]: Why is this removed? As we discussed offline, we'll continue to have unique id for various purpose in node.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp [hunk 2]
```diff
@@ -34,7 +34,6 @@ void remove_redundant_reorders::run(program& p) {
         if (!update_implementations)
             return;
 
-        node.set_unique_id();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:37]: here too.

File: src/plugins/intel_gpu/src/graph/include/program_node.h [hunk 3]
```diff
@@ -91,7 +91,7 @@ struct program_node {
                 if (u->get_dependencies().size() <= dep_idx) {
                     continue;
                 }
-                if (u->get_dependency(dep_idx).get_unique_id() == unique_id) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/include/program_node.h:94]: Please do not remove unique id usage in program_node. Its unnecessary for this task.

File: src/plugins/intel_gpu/src/runtime/kernels_cache.hpp [hunk 4]
```diff
@@ -52,28 +52,16 @@ class kernels_cache {
         std::shared_ptr<kernel_string> kernel_strings;
         std::string id;
         bool dump_custom_program;
-        size_t hash_value;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.hpp:55]: Also we should continue to use hash from string.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/post_input_reorder.cpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/post_input_reorder.cpp]: rollback to remove unique_id. thanks.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp (hunk 2 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:37]: rollback it.

File: src/plugins/intel_gpu/src/graph/include/program_node.h (hunk 3 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/program_node.h]: rollback it.

File: src/plugins/intel_gpu/include/intel_gpu/graph/program.hpp [hunk 5]
```diff
@@ -239,16 +239,14 @@ struct program {
                              std::shared_ptr<InferenceEngine::CPUStreamsExecutor> task_executor,
                              bool is_internal);
     static void init_primitives();
-    kernel_id add_kernel(const std::shared_ptr<kernel_string>& kernel_sring);
-    kernel::ptr get_kernel(kernel_id id);
+    void add_kernel(kernel_impl_params& params, const std::shared_ptr<kernel_string>& kernel_sring);
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/graph/program.hpp:242]: And also the function arg is preferably declared as const for better compilation opt.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/post_input_reorder.cpp [hunk 6]
```diff
@@ -70,8 +70,8 @@ void post_input_reorder::run(program& p) {
                 node->set_output_layout(previous_layout, false);
                 reorder.set_selected_impl(reorder.type()->choose_impl(reorder));
                 if (auto impl = reorder.get_selected_impl()) {
-                    auto kernel_ids = p.get_kernels_cache().add_kernels_source(impl->get_kernels_source());
-                    impl->set_kernel_ids(kernel_ids);
+                    auto params = reorder.get_kernel_impl_params();
+                    p.get_kernels_cache().add_kernels_source(*params, impl->get_kernels_source());
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/post_input_reorder.cpp:74]: here too.

File: src/plugins/intel_gpu/src/graph/impls/common/condition.cpp [hunk 7]
```diff
@@ -54,7 +54,7 @@ struct condition_impl : typed_primitive_impl<condition> {
         return make_unique<condition_impl>(arg);
     }
 
-    void init_kernels(const kernels_cache&) override {}
+    void init_kernels(const kernels_cache& , kernel_impl_params&) override {}
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/common/condition.cpp:57]: const params

File: src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp [hunk 8]
```diff
@@ -29,7 +29,7 @@ class wait_for_events_impl : public primitive_impl {
         return make_unique<wait_for_events_impl>(*this);
     }
 
-    void init_kernels(const kernels_cache&) override {}
+    void init_kernels(const kernels_cache&, kernel_impl_params&) override {}
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp:32]: const params

File: src/plugins/intel_gpu/src/graph/impls/ocl/custom_primitive.cpp [hunk 9]
```diff
@@ -50,11 +48,14 @@ struct custom_gpu_primitive_impl : typed_primitive_impl<custom_gpu_primitive> {
                              std::shared_ptr<kernel_selector::cl_kernel_data>& cl_kernel)
         : cl_kernel(cl_kernel)
         , _kernels() {
-        _kernel_id = arg.get_program().add_kernel(cl_kernel->code.kernelString);
+        auto _params = arg.get_kernel_impl_params();
+        arg.get_program().add_kernel(*_params, cl_kernel->code.kernelString);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/ocl/custom_primitive.cpp:52]: Why should we do add_kernel here? 
Can't custom primitives be handled like others, 
1) Set kernel string to impl here 
2) Before build_all : add_kernel

File: src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp [hunk 10]
```diff
@@ -38,21 +36,22 @@ struct generic_layer_impl : typed_primitive_impl<generic_layer> {
     generic_layer_impl(const generic_layer_node& arg)
         : _cl_kernel_data(*arg.get_primitive()->generic_params.clKernel.get())
         , _kernels() {
-        _kernel_id = arg.get_program().add_kernel(arg.get_primitive()->generic_params.clKernel->code.kernelString);
+        auto _params = arg.get_kernel_impl_params();
+        arg.get_program().add_kernel(*_params, arg.get_primitive()->generic_params.clKernel->code.kernelString);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp:40]: Same question as custom kernel.

File: src/plugins/intel_gpu/src/graph/include/generic_layer.hpp [hunk 11]
```diff
@@ -37,11 +37,15 @@ struct generic_layer : public primitive_base<generic_layer> {
     const kernel_selector::generic_kernel_params generic_params;
 
     size_t hash() const override {
-        size_t seed = primitive::hash();
+        size_t seed = 0;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/include/generic_layer.hpp:40]: Why should we change the basic seed to 0?

File: src/plugins/intel_gpu/src/graph/include/generic_layer.hpp [hunk 12]
```diff
@@ -37,11 +37,15 @@ struct generic_layer : public primitive_base<generic_layer> {
...
         seed = hash_combine(seed, id);
         return seed;
     }
 
+    bool operator==(const primitive& rhs) const override {
+        return (id == rhs.id);
+    }
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/include/generic_layer.hpp:47]: It was already using id as hash. Why should we change this?

File: src/plugins/intel_gpu/src/graph/kernel_impl_params.cpp [hunk 13]
```diff
@@ -15,7 +15,9 @@
 namespace cldnn {
 
 size_t kernel_impl_params::hash() const {
-    size_t seed = desc->hash();
+    size_t seed = 0;
+    if (desc != nullptr)
+        seed = desc->hash();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/kernel_impl_params.cpp:20]: When desc is null?

Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/kernel_impl_params.cpp:20]: For serialization?

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 14]
```diff
@@ -90,48 +89,58 @@ void kernels_cache::get_program_source(const kernels_code& kernels_source_code,
...
+                current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
+            }
+
+            // Create new kernels batch when the limit is reached
+            // and current kernel's entry_point is duplicated in this kernels batch
+            if (current_bucket.back().kernels_counter >= get_max_kernels_per_batch()
+                || std::find(entry_points.begin(), entry_points.end(), entry_point) != entry_points.end()) {
+                const auto& batch_id = static_cast<int32_t>(current_bucket.size());
+                current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
+                entry_points.clear();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:132]: Where is the duplication of kernels handled? e.g., two kernels with same impl_param, same entry. 
Should't we distribute batches like 
1. scan the entry point and check dupilication 
- [entry_point_A] : {kernel_from impl_param_A_1), kernel_from_impl_param_A_2 }
- [entry_point_B] : kernel_from_impl_param_B_1
- ...  
2. Distribute kernel_from_impl_param_A_1 and kernel_from_impl_param_A_2 to different batches 
- min # of batches : max # of bucket size of above.

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:132]: I think duplication checking code is missed during code rebase. I will update it.

File: src/plugins/intel_gpu/include/intel_gpu/graph/program.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/graph/program.hpp]: Fixed it.

File: src/plugins/intel_gpu/include/intel_gpu/primitives/custom_gpu_primitive.hpp [hunk 15]
```diff
@@ -75,21 +75,13 @@ struct custom_gpu_primitive : public primitive_base<custom_gpu_primitive> {
     const primitive_id_arr kernels_code;
 
     size_t hash() const override {
-        size_t seed = primitive::hash();
-        seed = hash_combine(seed, kernel_entry_point);
-        seed = hash_combine(seed, kernels_code.size());
+        size_t seed = 0;
+        seed = hash_combine(seed, id);
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/custom_gpu_primitive.hpp]: Fix hash and ==operator function

File: src/plugins/intel_gpu/src/graph/impls/common/condition.cpp (hunk 7 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/common/condition.cpp]: Fixed it.

File: src/plugins/intel_gpu/src/graph/impls/common/loop.cpp [hunk 16]
```diff
@@ -23,7 +23,7 @@ struct loop_impl : typed_primitive_impl<loop> {
         return make_unique<loop_impl>(*this);
     }
 
-    void init_kernels(const kernels_cache&) override {}
+    void init_kernels(const kernels_cache& , kernel_impl_params&) override {}
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/common/loop.cpp]: Fixed it.

File: src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp (hunk 8 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp]: Fixed it.

File: src/plugins/intel_gpu/src/graph/include/generic_layer.hpp (hunk 11 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/generic_layer.hpp]: Modify hash and ==operator to use genric_params instead of id and remove zero seed

File: src/plugins/intel_gpu/src/graph/include/generic_layer.hpp (hunk 12 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/generic_layer.hpp:135]: Modify code to use generic_params instead of id.

File: src/plugins/intel_gpu/src/graph/impls/ocl/custom_primitive.cpp (hunk 9 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/ocl/custom_primitive.cpp]: Fixed it.

File: src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp (hunk 10 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/ocl/generic_layer.cpp]: Fixed it

File: src/plugins/intel_gpu/include/intel_gpu/graph/serialization/pair_serializer.hpp [hunk 17]
```diff
@@ -0,0 +1,32 @@
+// Copyright (C) 2018-2023 Intel Corporation
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/graph/serialization/pair_serializer.hpp:1]: nit: just 2023 for new file

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 18]
```diff
@@ -515,10 +515,11 @@ network::~network() {
 //     [ executable primitive_inst ]
 //     [ memory reuse information ]
 void network::save(cldnn::BinaryOutputBuffer& ob) {
-    kernels_cache kernels_cache(get_engine(), _config, 0, nullptr, {""});
+    kernels_cache kernels_cache(get_engine(), _config, 0);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/network.cpp:518]: BTW I recognized this was from the original implementation, but would like to ask just out of curiosity.
Can't we use program's kernels cache as is?

File: src/plugins/intel_gpu/include/intel_gpu/graph/serialization/pair_serializer.hpp (hunk 17 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/graph/serialization/pair_serializer.hpp]: Fixed it. thanks

File: src/plugins/intel_gpu/src/runtime/kernels_cache.hpp [hunk 19]
```diff
@@ -22,9 +23,31 @@
 #include "ocl/ocl_engine.hpp"
 
 namespace cldnn {
+using cached_kernel_id_type = std::pair<size_t, std::string>;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.hpp:26]: As a note: The only thing I have in mind as a concern is the potential hash collision of the key, 
because not only "entry_key" but also the hash from the binary. 
From the offline conversation with eddy, he can replace the key without binary hash.

File: src/plugins/intel_gpu/src/graph/network.cpp (hunk 18 above)
Comment (e-ddykim) [reply src/plugins/intel_gpu/src/graph/network.cpp:518]: I updated to reuse the program's kernels cache as you reviewed. Thank you.

File: src/plugins/intel_gpu/src/runtime/kernels_cache.hpp (hunk 19 above)
Comment (e-ddykim) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.hpp:26]: I updated not to use hash. Thank you.

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.hpp]: @e-ddykim , @yeonbok Updated commits are merged to this branch now. thanks.

File: src/plugins/intel_gpu/src/graph/impls/ocl/kernel_selector_helper.cpp [hunk 20]
```diff
@@ -101,12 +101,13 @@ bool query_local_block_io_supported(engine& e, const ExecutionConfig& config) {
     kernel_string->batch_compilation = true;
 
     try {
+        kernel_impl_params dummpy_params;
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/kernel_selector_helper.cpp:104]: nit: `dummy_params`

File: src/plugins/intel_gpu/src/graph/kernel_impl_params.cpp [hunk 21]
```diff
@@ -32,7 +34,7 @@ size_t kernel_impl_params::hash() const {
 }
 
 bool kernel_impl_params::operator==(const kernel_impl_params& rhs) const {
-    if (*desc != *rhs.desc)
+    if (desc != nullptr && rhs.desc != nullptr && *desc != *rhs.desc)
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/kernel_impl_params.cpp:37]: what if one desc is nullptr and another is not?

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/kernel_impl_params.cpp]: Fixed it.

File: src/plugins/intel_gpu/src/graph/impls/ocl/kernel_selector_helper.cpp (hunk 20 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/ocl/kernel_selector_helper.cpp]: Fixed it. thanks.

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 22]
```diff
@@ -328,14 +334,31 @@ void kernels_cache::build_batch(const engine& build_engine, const batch_program&
...
-    if (_kernels.end() == res)
-        throw std::runtime_error("Kernel " + id + " not found in the kernel cache!");
-    return res->second;
+    auto res = _kernels.find(params);
+    if (_kernels.end() == res) {
+        std::string issued_id;
+        if (params.desc) {
+            issued_id = params.desc->id;
+        }
+        throw std::runtime_error("Kernel for {" + issued_id + "} is not found in the kernel cache!");
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.cpp]: Fixed it.

File: src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp [hunk 23]
```diff
@@ -529,9 +529,11 @@ struct layout {
         seed = hash_combine(seed, format.value);
         seed = hash_combine(seed, data_type);
 
-        auto pshape = get_partial_shape();
-        for (size_t idx = 0; idx < pshape.size(); idx++) {
-            seed = hash_combine(seed, pshape[idx].get_length());
+        if (!is_dynamic()) {
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp]: Fixed it.

[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp]: Fixed it.

File: src/plugins/intel_gpu/include/intel_gpu/runtime/utils.hpp [hunk 24]
```diff
@@ -215,6 +215,12 @@ static size_t hash_range(size_t seed, It first, It last) {
     return seed;
 }
 
+template <typename T>
+static bool is_equal(const std::vector<T>& vec1, const std::vector<T>& vec2) {
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/utils.hpp]: Fixed it.

**Review by yeonbok — APPROVED** [review]

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 25]
```diff
@@ -430,117 +453,110 @@ void kernels_cache::reset() {
...
 }
 
-void kernels_cache::add_kernels(const std::vector<std::string>& kernel_ids, const std::vector<kernel::ptr>& kernels) {
-    OPENVINO_ASSERT(kernel_ids.size() == kernels.size(), "[GPU] The sizes of kernel_ids and kernels are different.");
+std::string kernels_cache::get_cached_kernel_id(kernel::ptr kernel) const {
+    auto ocl_kernel = std::static_pointer_cast<cldnn::ocl::ocl_kernel>(kernel);
+    const auto& entry_point = ocl_kernel->get_handle().getInfo<CL_KERNEL_FUNCTION_NAME>();
+    auto program = ocl_kernel->get_handle().getInfo<CL_KERNEL_PROGRAM>();
+    cl::vector<unsigned char> program_binaries = getProgramBinaries(program);
+    auto iter = _cached_binaries.find(program_binaries);
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:474]: In get_cached_kernel_id, _cached_bineries is not modified. it is just used to check whether the binrary is existed in _cached_binaries.
_cached_binaries is only modified in add_to_cached_kernels and add_to_cached_kernels has already mutex lock.
(After I added mutex lock in this function, cache kernel functional test has fallen into a deadlock)

**Review by e-ddykim — APPROVED** [review]
Looks good to me
